batch/batch/cloud/azure/resources.py

```diff
@@ -0,0 +1,228 @@
+import abc
+from typing import Dict, Any, Optional
+
+from ...driver.billing_manager import ResourceVersions
+from ...resources import (QuantifiedResource, Resource, DiskResourceMixin, VMResourceMixin, IPFeeResourceMixin,
+                          ServiceFeeResourceMixin, ExternalDiskResourceMixin)
+from .resource_utils import azure_disk_from_storage_in_gib, azure_disks_by_disk_type
+
+
+class AzureResource(Resource, abc.ABC):
+    pass
+
+
+def azure_disk_prefix(disk_product: str, redundancy_type: str, location: str) -> str:
+    return f'az/disk/{disk_product}_{redundancy_type}/{location}'
+
+
+class AzureDiskResource(DiskResourceMixin, AzureResource):
+    FORMAT_VERSION = 1
+    TYPE = 'azure_disk'
+
+    @staticmethod
+    def generate_prefix(disk_product: str, redundancy_type: str, location: str) -> str:
+        return azure_disk_prefix(disk_product, redundancy_type, location)
+
+    @staticmethod
+    def from_dict(data: Dict[str, Any]) -> 'AzureDiskResource':
+        assert data['type'] == AzureDiskResource.TYPE
+        return AzureDiskResource(data['name'], data['storage_in_gib'])
+
+    @staticmethod
+    def new_resource(resource_versions: ResourceVersions,
+                     disk_type: str,
+                     storage_in_gib: int,
+                     location: str,
+                     redundancy_type: str = 'LRS',
+                     ) -> 'AzureDiskResource':
+        assert redundancy_type in ('LRS', 'ZRS'), redundancy_type
+
+        # Azure bills for specific disk sizes so we must round the storage_in_gib to the nearest power of two
+        disk = azure_disk_from_storage_in_gib(disk_type, storage_in_gib)
+        assert disk, f'disk_type={disk_type} storage_in_gib={storage_in_gib}'
+
+        prefix = AzureDiskResource.generate_prefix(disk.name, redundancy_type, location)
```

Comment:
I say this below, but I again find the term "prefix" obfuscating here. This is really the name of the kind/type of resource with the given disk type, redundancy type, and location. Then, on the next line, we say "get me the latest *fully versioned* resource of this name".

"generate_prefix" gives me the "kind" or "type" of resource which fits my needs. "latest_resource_name" finds the latest "version" of that "kind/type" of resource.

---

Hmm.

OK, here's a question: why are we concatenating the version into the resource's name in the database? That feels like the opposite of how databases are supposed to work. Each distinct piece of information should be in a column.

This feels more in the spirit of SQL and also less obfuscating of the difference between resource types and resource versions:

```sql
CREATE TABLE IF NOT EXISTS `resources` (
  `name` VARCHAR(100) NOT NULL,
  `version` VARCHAR(100) NOT NULL,
  `rate` DOUBLE NOT NULL,
   PRIMARY KEY (`name`, `version`)
 ) ENGINE = InnoDB;

 CREATE TABLE IF NOT EXISTS `latest_resource_versions` (
   `name` VARCHAR(100) NOT NULL,
   `version` VARCHAR(100) NOT NULL,
   PRIMARY KEY (`name`)
 ) ENGINE = InnoDB;
```

We should probably make name a `UNIQUE KEY`. That will ensure there's always exactly one latest version. We couldn't achieve that if we used a boolean flag on the resources table; therefore, I do like having two tables.

Now we can pervasively use the term "name" to refer to the kind/type of resource and "version" to refer to the version of it (which in the Azure case is always a date stamp).

Reply:
We might even want a `VersionedResource` which is a pair of a version and a name. We could have a `VersionedRatedResource` which has a version, rate, and a name.

Reply:
I agree with you that when we created the database resources table, we should have added a version field. If we change the table structure, there's a couple of things that I worry about:

1. Backwards compatibility with old workers. This should be easy to fix
2. We'll have to figure out how to split the string in the resources table to get rid of the version (which is always 1 right now) which is not straightforward.
3. We'll also need to account for the aggregated_*_resources tables.

In all, I thought it was easier to do what I did and have the full name still in the database.

Reply:
There actually is a function that we can use: `SUBSTRING_INDEX`

Reply:
If you are okay with a huge migration (we should backup the database ahead of time!!!!) and we'll need to take Batch offline for a long time, then we can add the proper tables.